### DISCUSSION TITLE
Connexion between my django API and apache airflow

### DISCUSSION THREAD

**Discussion by tcunique** [apache/airflow#29965] [Q&A] [answered]
Hello ! 

First of all excuse me if my english is not great, it is not my native language so I'll try my best to be clear enough.
I am trying to connect a django API with apache airflow and to check if the connexion was made, i wrote a dag named django_dag.py. 
But, when I start the dag, i get this log : 

<img width="1276" alt="Capture d’écran 2023-03-07 à 22 11 01" src="https://user-images.githubusercontent.com/114347777/223553626-ae606ea9-0fbc-489a-bd85-2fc5528b0c0b.png">

My django api is correctly launch as we can see below : 

<img width="665" alt="Capture d’écran 2023-03-07 à 22 11 46" src="https://user-images.githubusercontent.com/114347777/223553782-2762a8b4-000a-488f-b8f2-950b2b3af9d4.png">

And here is the connexion that i've filled in apache airflow :

<img width="987" alt="Capture d’écran 2023-03-07 à 22 12 35" src="https://user-images.githubusercontent.com/114347777/223553964-c508db2b-8938-4430-9021-6dfee754e88b.png">

Thank you in advance for the time you've took to read my message

Thomas

**Comment by pierrejeambrun** [comment]
Hello @tcunique,

Does your worker have access to this API ? It looks like it is hosted locally, depending on how your worker is executed and on which network (docker for instance) it needs to be able to call your API. You can try with a publicly available API first.

Also we don't have your DAG code, so it is not easy to see what you are trying to do. By default PythonOperator does not leverage http connections.

You can take a look at the `SimpleHttpOperator` which serve exactly this purpose (making http calls).
```python
    SimpleHttpOperator(
        task_id="http_get",
        endpoint="endpoint_path",
        method="get",
        http_conn_id="django_api",
        headers={"Content-Type": "application/json"},
    )
```

If you want to incorporate that directly in your PythonOperator, because it performs other stuff as well you can use the HttpHook, a minimal example is:
```python
def python_callable() -> None:
    # Do some stuff...
    response = HttpHook("get", http_conn_id="django_api").run(
        endpoint="get", headers={"Content-Type": "application/json"}
    )
    print(response.json())
    # Do some other stuff...

# Note, on the public API I am requesting for testing this, the endpoint is "get" i.e https://somedomain/get.
# This is just a coincidence should not be mistaken with the method passed to the HttpHook param.
    
# In your DAG
PythonOperator(
    task_id="http_get_with_python_operator",
    python_callable=python_callable,
)
```

**Comment by tcunique** [comment]
Hello ! 
Thank you for your response ! 

My API is hosted locally yes, with docker.
I've tried to put the simpleHttpOperator but unfortunately it still gives me the same error.



```py
from airflow import DAG
from airflow.hooks.base import BaseHook
import requests
from requests.auth import HTTPBasicAuth
from airflow.providers.http.operators.http import SimpleHttpOperator
import datetime

default_args = {
    'owner': 'airflow',
    'retries': 5,
    'retry_delay': datetime.timedelta(minutes=5)
}

def verify_api_connection():
    conn_id = 'django_api'
    conn = BaseHook.get_connection(conn_id)
    url = f"{conn.host}:{conn.port}/api/health_check/"
    response = requests.get(url, auth=HTTPBasicAuth(conn.login, conn.password))
    if response.status_code == 200:
        print("API connection successful!")
    else:
        print("API connection failed.")

with DAG(
    dag_id="dag_with_api", 
    default_args=default_args,
    start_date=datetime.datetime(2023, 3, 2, 15, 30, 0),
    schedule_interval="@daily",
) as dag:

    task1 = SimpleHttpOperator(
        task_id="http_get",
        endpoint="endpoint_path",
        method="get",
        http_conn_id="django_api",
        headers={"Content-Type": "application/json"},
    )

    task1
```

Thank you for your time ! 

Thomas

**Comment by pierrejeambrun** [comment, marked as answer]
Hello,

Try with a publicly available api. (Not locally hosted on docker) and this will work.

Your connection is wrong. From the worker container, localhost:8000 is not where your api is running. (The api is not running inside the worker container)

You can take a look at docker networking and how to reach another container from a container (service:port etc).

**Comment by tcunique** [comment]
Yes, it finally worked with a public API, thank you so much, have a great day !
